File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        stack = []
        lookup = {')': '(', ']': '[', '}': '{'}

        if not s:
            return False

        for i in s:
            if i in lookup.values():
                stack.append(i)
            elif stack and lookup[i] == stack[-1]:
                stack.pop()
            else:
                return False
            
        return stack == []
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {

            if (c == '{' || c == '(' || c == '[')
                stack.push(c);
            else {
                if (stack.isEmpty()) 
                    return false;
                else {
                    if (stack.peek() == '{' && c == '}' ||
                            stack.peek() == '[' && c == ']' ||
                            stack.peek() == '(' && c == ')')
                        stack.pop();
                    else
                        return false;
                }
            }
        }
        return stack.isEmpty();
    }
}